src/utils/generic.rs

```diff
@@ -11,28 +11,41 @@ use core::{
     },
 };
 
-use blake2::Blake2b;
+use blake2::{crypto_mac::Mac, Blake2b};
 use curve25519_dalek::scalar::Scalar;
 
 use crate::{errors::ProofError, protocols::scalar_protocol::ScalarProtocol, range_proof::RangeProof};
 
 /// Create a Blake2B deterministic nonce given a seed, label and index
-pub fn nonce(seed_nonce: &Scalar, label: &str, index: Option<usize>) -> Result<Scalar, ProofError> {
+pub fn nonce(
+    seed_nonce: &Scalar,
+    label: &str,
+    index_j: Option<usize>,
+    index_k: Option<usize>,
+) -> Result<Scalar, ProofError> {
     let encoded_label = label.as_bytes();
     if encoded_label.len() > 16 {
         // See https://www.blake2.net/blake2.pdf section 2.8
         return Err(ProofError::InvalidLength("nonce label".to_string()));
     };
-    let hasher = if let Some(salt) = index {
-        let encoded_index = salt.to_le_bytes();
-        if encoded_index.len() > 16 {
+    let mut hasher = if let Some(salt) = index_j {
+        let encoded_salt = salt.to_le_bytes();
+        if encoded_salt.len() > 16 {
             // See https://www.blake2.net/blake2.pdf section 2.8
-            return Err(ProofError::InvalidLength("nonce index".to_string()));
+            return Err(ProofError::InvalidLength("nonce index_j".to_string()));
         };
-        Blake2b::with_params(&seed_nonce.to_bytes(), &encoded_index, encoded_label)
+        Blake2b::with_params(&seed_nonce.to_bytes(), &encoded_salt, encoded_label)
     } else {
         Blake2b::with_params(&seed_nonce.to_bytes(), &[], encoded_label)
     };
+    if let Some(salt) = index_k {
+        let encoded_salt = salt.to_le_bytes();
+        if encoded_salt.len() > 16 {
+            // See https://www.blake2.net/blake2.pdf section 2.8
+            return Err(ProofError::InvalidLength("nonce index_k".to_string()));
+        };
+        hasher.update(&encoded_salt);
+    }
 
     Ok(Scalar::from_hasher_blake2b(hasher))
```

Comment:
Has it been verified that the hash output with no `k` set (which should hash the empty string) is in fact unique across different nonces? It should be, but it might be nice to test this specifically to ensure the hasher is working as expected when no input is passed to it.

Reply:
Yes, that is specifically tested.

Reply:
Added additional tests